### ISSUE TITLE
PageTemplate handling of `default` broken starting with release 4.4

### ISSUE THREAD

**Issue by dataflake** [zopefoundation/Zope#846]
Symptom: Attempting to use the `default` keyword in a TAL expression to render the default contents of a tag will not render that content, but something like `<object object at 0x105810e30>`. This happens regardless of how `default` is used:

```
<p tal:define="foo string:" tal:content="python: foo or default">Defaul content</p>
```
 or

```
<p tal:content="context/unknownvariable|default">Defaul content</p>
```

Tagging Dieter because 4.4 contained major PageTemplate-related changes from him.

**Comment by dataflake** [comment]
P.S.: The issue is in Zope itself, not the dependencies. If I run Zope 4.3 with the latest dependencies from the 4.x branch the issue does not occur.

**Comment by d-maurer** [comment]
Jens Vagelpohl wrote at 2020-6-5 04:45 -0700:
>Symptom: Attempting to use the `default` keyword in a TAL expression to render the default contents of a tag will not render that content, but something like `<object object at 0x105810e30>`. This happens regardless of how `default` is used:

The problem is understood.

`chameleon` generates code of the following form for a template:
```
__filename = '<string>'
__default = object()
__tokens = {18: (u'default', 1, 18)}
...
def initialize(modules, nothing, tales, zope_version_4_4_3_):

    def render(__stream, econtext, rcontext, __i18n_domain=None, __i18n_context=None):
	...
    return {'render': render, }
```

When the template is rendered, it calls the `render` function
returned by a call to the `initialize` above.

Older `chameleon` versions likely have used the `__default`
definition from the start of the generated code (otherwise, the purpose
of the definition would be unclear).
Modern `chameleon` versions, however, no longer use this definition;
instead they represent the TALES `default` value
by a string constant locally defined in `render`.

Thus, the TALES `default` value is not represented by a unique
object. Instead, each template has its own private representation.
I consider this a (`chameleon`) bug: if template "A" calls template "B"
and passes its `default` as parameter, template "B" will not understand
it.


It will not be easy to fix this in the integration.

I will explore the following approach:
When a `Products.PageTemplates.PageTemplate.PageTemplate`
is rendered, it determines the `render` function used
by the corresponding `chameleon` template (this needs to
use some implementation details). It looks whether it finds
a single `'__default'` contant in this function and in this
case uses it as its representation of the TALES `default` object.

This assumes that Python maps multiple occurrences
of the string contant `'__default'` inside the same function
to the same object. To my knowledge, Python does not garantee this.
However, apparently, `chameleon`, too, depends on this assumption.


Maybe, @malthe may come up with a cleaner solution in `chameleon`.

**Comment by malthe** [comment]
@d-maurer can you check if this solves your issue? https://github.com/malthe/chameleon/pull/313

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-6-7 14:18 -0700:
>@d-maurer can you check if this solves your issue? https://github.com/malthe/chameleon/pull/313

I believe that the `__default` defined in the prelude of the
generated template code is no longer used by the (generated) `render` function:
I had replaced that definition in the integration (into `Zope`)
by `from zope.tales.tales import _default as __default` and
nevertheless @dataflake observed wrong `default` handling
(and when I looked at the generated code, I saw `is` comparisons
against the string `'__default'` and not against the object `__default`);
which means that this definition is irrelevant.

If the `__default` defined in the prelude is used again
(by the (generated) `render`), my original integration
approach would work - effectively forcing `chameleon` to
use the `zope.tales` `default` object. Of course, it was
a hack in the first place and I would not be unhappy to
get rid of it.

If the `__default` defined in the prelude is used
and defined as `intern('__default')` I could use that
knowledge in the integration (using that all `intern(X)`
whereever they occur always refer to the same object (dependent on X)).
But, of course, I would prefer not to use this implementation
detail.


What about the following approach:

 * you define somewhere in `chameleon`, say in `chameleon.tal`,
   an object `default`
   (the definition could look like
   `default = '__default'`
   or `default = intern('__default')`
   or `default = object()` or whatever).

 * the prelude contains
   `from chameleon.tal import default as __default`

 * the (generated) `render` uses the `__default` defined
   in the prelude for `default` checking (and not the
   string `'__default'`)

 * I use `chameleon.tal.default` as the TALES `default`
   object in the integration


I just remembered a `chameleon` problem related to its persistent caching:
whenever the generated code changes in a way important for
the external world (like the above sketched change for the
TALES `default`), cached code does not contain the change
and may break (if the external world depends on the change).
Therefore, I suggest to include version information in the generated
cache key to ensure that with any important version change the
cache content is automatically invalidated.

**Comment by dataflake** [comment]
> @d-maurer can you check if this solves your issue? [malthe/chameleon#313](https://github.com/malthe/chameleon/pull/313)

For what it's worth, the change does not fix the issue on current Zope 4.x without Dieter's additional changes.

**Comment by malthe** [comment]
Version-information is indeed part of the generated cache key.

I did consider doing `from chameleon.tales import default` but then I thought that it was a bit odd given that we don't important any other things besides from the standard library, at least as far as I could tell.

I mean the value of `default` really should almost be an implementation detail (at least the default value of the default marker).

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-6-7 23:22 -0700:
>Version-information is indeed part of the generated cache key.

I may have overlooked that.


>I did consider doing `from chameleon.tales import default` but then I thought that it was a bit odd given that we don't important any other things besides from the standard library, at least as far as I could tell.
>
>I mean the value of `default` really should almost be an implementation detail (at least the default value of the default marker).

In the `Zope` world TALES is used by many components. Some examples
are `CMFCore`, `DCWorkflow`, some Plone portlets.
All of those are using the `zope.tales` implementation.

`zope.tales` is much easier to integrate and extend than `chameleon.tales`:
with `zope.tales`, an expression is implemented as a function
called with a context to produce a value; with `chameleon.tales` as
a function which generates code which when executed in an environment
produces the value.

I believe that a single TALES implementation
should be used in the Zope world and it should be `zope.tales` (being
the one easier to use and extend).

But the TALES implementation must necessarily know the `default` marker:
this marker cannot be an implementation detail; there must be something
which connects the TALES `default` value with the template engine
to ensure that the engine takes the wanted action.


If your cache keys contain version information, why is it important
not to import things in the generated code from `chameleon`?.
`chameleon` is available and the version information ensures that it
is an appropriate version. Thus, it should be safe to import from there.

**Comment by malthe** [comment]
It could also be an argument that is provided to the render function and drawn from the the template class. I think that's a nicer pattern which then also allows for overriding the value.

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-6-7 23:22 -0700:
> ...
>I mean the value of `default` really should almost be an implementation detail (at least the default value of the default marker).

Thinking more about it, you likely could make it an implementation
detail. At the moment, this seems not yet to be the case.

A way to make it an implementation detail could be to put
`default` into the rendering context (i.e. `Scope`) at rendering start.
This way, the TALES `default` would be mapped to whatever the
template engine uses as "default marker".

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-6-8 00:06 -0700:
>It could also be an argument that is provided to the render function and drawn from the the template class. I think that's a nicer pattern which then also allows for overriding the value.

I would be fine with this, too.

**Comment by dataflake** [comment]
The fix is in the latest release, Zope 4.4.4, as well as on the `master` branch
